### PR TITLE
Migrate forest to new cursor API

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Migrates forest to new cursor API.

This mainly impacts object-forest (which has to implement the new API), and editable-tree (which has to consume the new API).

Other than migrating to the new API, there are some significant design changes in object-forest:
1. It now uses MapTree instead of JsonableTree
2. It has a dummy root node above the detached sequences which simplifies apply delta and some other aspects of root handling.
3. Its cursor wraps an existing cursor instead of extending one.

## Performance:
Faster, though one regression:

Before
```
status  name                                                     period (ns/op)  relative margin of error  iterations  samples  total time (s)
------  -------------------------------------------------------  --------------  ------------------------  ----------  -------  --------------
    ✔   cursorToJsonObject(object-forest Cursor): 'canada'         45,195,222.5                    ±1.81%          82       41            5.79
    ✔   jsonableTreeFromCursor(object-forest Cursor): 'canada'     96,763,088.4                    ±3.05%          30       30            5.94
    ✔   sum(object-forest Cursor): 'canada'                        48,716,704.3                    ±0.22%          76       38            5.84
    ✔   averageTwoValues(object-forest Cursor): 'canada'           36,658,514.6                    ±1.54%          98       49            5.64
    ✔   cursorToJsonObject(object-forest Cursor): 'twitter'        41,951,789.8                    ±1.16%          86       43            5.74
    ✔   jsonableTreeFromCursor(object-forest Cursor): 'twitter'    80,564,255.5                    ±3.16%          35       35            5.77
    ✔   sum(object-forest Cursor): 'twitter'                       38,948,035.3                    ±0.13%          92       46            5.67
    ✔   averageTwoValues(object-forest Cursor): 'twitter'             302,512.5                    ±0.94%      16,245       95            5.43
```

After:
```
status  name                                                     period (ns/op)  relative margin of error  iterations  samples  total time (s)
------  -------------------------------------------------------  --------------  ------------------------  ----------  -------  --------------
    ✔   cursorToJsonObject(object-forest Cursor): 'canada'         53,977,035.4                    ±1.02%          49       49            5.59
    ✔   jsonableTreeFromCursor(object-forest Cursor): 'canada'     83,449,600.6                    ±3.66%          34       34            5.86
    ✔   sum(object-forest Cursor): 'canada'                        27,261,029.8                    ±0.27%         126       63            5.49
    ✔   averageTwoValues(object-forest Cursor): 'canada'           21,406,211.2                    ±0.27%         183       61            5.63
    ✔   cursorToJsonObject(object-forest Cursor): 'twitter'        32,734,807.8                    ±1.26%         108       54            5.61
    ✔   jsonableTreeFromCursor(object-forest Cursor): 'twitter'    57,881,098.4                    ±2.75%          46       46            5.63
    ✔   sum(object-forest Cursor): 'twitter'                       16,481,119.6                    ±0.52%         252       63            5.50
    ✔   averageTwoValues(object-forest Cursor): 'twitter'             241,298.7                    ±0.45%      20,657       91            5.41

```

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 1]
```diff
@@ -20,27 +19,39 @@ import {
...
+import { mapTreeFromCursor, singleMapTreeCursor } from "../mapTreeCursor";
+
+function makeRoot(): MapTree {
+    return {
+        type: brand("above root placeholder"),
+        fields: new Map(),
+    };
+}
 
 export class ObjectForest extends SimpleDependee implements IEditableForest {
```
Comment (jenn-le) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:47]: Needs a doc comment

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 2]
```diff
@@ -74,9 +87,6 @@ export class ObjectForest extends SimpleDependee implements IEditableForest {
     }
 
     applyDelta(delta: Delta.Root): void {
```
Comment (jenn-le) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:89]: I think this warrants a doc comment as well.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 3]
```diff
@@ -351,32 +383,34 @@ class Cursor extends RootedTextCursor implements ITreeSubscriptionCursor {
...
-        this.keyStack.length = 0;
-        this.siblingStack.length = 0;
-        this.indexStack.length = 0;
-        this.index = -1;
+        this.innerCursor = undefined;
         this.forest.currentCursors.delete(this);
     }
 
-    public set(root: DetachedField, index: number): void {
+    public set(): void {
```
Comment (jenn-le) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:390]: I think this also needs a comment. Hard to tell set means when reading through the ObjectForest code.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 4]
```diff
@@ -87,24 +97,17 @@ export class ObjectForest extends SimpleDependee implements IEditableForest {
...
-                dstField = getGenericTreeField(cursor.getNode(), currentField, true);
+            const detachedKey = detachedFieldAsKey(toAttach);
+            const children = getMapTreeField(this.roots, detachedKey, false);
+            this.roots.fields.delete(detachedKey);
+            if (children.length === 0) {
+                return 0; // Prevent creating 0 sized fields when inserting empty into empty.
             }
+
+            const [parent, key] = cursor.getParent();
+            const dstField = getMapTreeField(parent, key, true);
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: Thats mostly an old habit on my part. Probably shouldn't do it. I'm more ok with that kind of think when its local and with common abbreviations, but I should fix it.

Also, worth noting that the use of "Id" in this file is probably worse: it's an abbreviation for both identifier and identity, and we don't know which it is! I'm going to leave that for now, but someday we should fix it.

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: Fixed.

[the remark this answers is not in the document]
Comment (jenn-le) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: Oh that's a good point. I'm so used to id that it didn't even register.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:53]: Done. Annoying this is mostly rehashing the folder level readme which we can't easily link from here.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:95]: This is part of the EditableForest API and thus inherits documentation from the interface.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: done. Also renamed.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 5]
```diff
@@ -351,32 +389,38 @@ class Cursor extends RootedTextCursor implements ITreeSubscriptionCursor {
...
-        this.siblingStack.length = 0;
-        this.indexStack.length = 0;
-        this.index = -1;
+        this.innerCursor = undefined;
         this.forest.currentCursors.delete(this);
     }
 
-    public set(root: DetachedField, index: number): void {
+    /**
+     * Move this cursor the special dummy node above the detached sequences.
```
Comment (jenn-le) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:397]:
```suggestion
     * Move this cursor to the special dummy node above the detached sequences.
```

**Review by jenn-le — APPROVED** [review]

**Pull request merged**